File: wallet/src/main/java/info/blockchain/wallet/payload/BalanceManager.java
package info.blockchain.wallet.payload;

import java.math.BigInteger;
import java.util.HashMap;
import java.util.List;
import java.util.Set;

import javax.annotation.Nonnull;

import info.blockchain.api.blockexplorer.BlockExplorer;
import info.blockchain.api.blockexplorer.FilterType;
import info.blockchain.api.data.Balance;
import info.blockchain.balance.CryptoCurrency;
import info.blockchain.balance.CryptoValue;
import retrofit2.Call;

public abstract class BalanceManager {

    private final BlockExplorer blockExplorer;
    private final CryptoCurrency cryptoCurrency;

    @Nonnull
    private CryptoBalanceMap balanceMap;

    BalanceManager(@Nonnull BlockExplorer blockExplorer, @Nonnull CryptoCurrency cryptoCurrency) {
        this.blockExplorer = blockExplorer;
        this.cryptoCurrency = cryptoCurrency;
        balanceMap = CryptoBalanceMap.zero(cryptoCurrency);
    }

    public void subtractAmountFromAddressBalance(String address, BigInteger amount) {
        balanceMap = balanceMap.subtractAmountFromAddress(address, new CryptoValue(cryptoCurrency, amount));
    }

    @Nonnull
    public BigInteger getAddressBalance(String address) {
        return balanceMap.get(address).toBigInteger();
    }

    @Nonnull
    BigInteger getWalletBalance() {
        return balanceMap.getTotalSpendable().toBigInteger();
    }

    @Nonnull
    public BigInteger getImportedAddressesBalance() {
        return balanceMap.getTotalSpendableImported().toBigInteger();
    }

    public void updateAllBalances(
        Set<String> xpubs,
        Set<String> importedAddresses
    ) {
        balanceMap = CryptoBalanceMapKt.calculateCryptoBalanceMap(
            cryptoCurrency,
            getBalanceQuery(),
            xpubs,
            importedAddresses
        );
    }

    private BalanceCall getBalanceQuery() {
        return new BalanceCall(blockExplorer, cryptoCurrency);
    }

    /**
     * @deprecated Use getBalanceQuery
     */
    @Deprecated
    public Call<HashMap<String, Balance>> getBalanceOfAddresses(List<String> addresses) {
        return getBlockExplorer().getBalance("btc", addresses, FilterType.RemoveUnspendable);
    }

    BlockExplorer getBlockExplorer() {
        return blockExplorer;
    }
}

File: wallet/src/main/java/info/blockchain/wallet/payload/BalanceManager.kt
package info.blockchain.wallet.payload

import com.blockchain.api.NonCustodialBitcoinService
import com.blockchain.api.bitcoin.data.BalanceResponseDto
import info.blockchain.balance.CryptoCurrency
import info.blockchain.balance.CryptoValue
import info.blockchain.wallet.payload.data.XPub
import info.blockchain.wallet.payload.data.XPubs
import retrofit2.Call

import java.math.BigInteger

abstract class BalanceManager constructor(
    val bitcoinApi: NonCustodialBitcoinService,
    private val cryptoCurrency: CryptoCurrency
) {
    private var balanceMap: CryptoBalanceMap

    val walletBalance: BigInteger
        get() = balanceMap.totalSpendable.toBigInteger()

    val importedAddressesBalance: BigInteger
        get() = balanceMap.totalSpendableImported.toBigInteger()

    private val balanceQuery: BalanceCall
        get() = BalanceCall(bitcoinApi, cryptoCurrency)

    init {
        balanceMap = CryptoBalanceMap.zero(cryptoCurrency)
    }

    fun subtractAmountFromAddressBalance(address: String, amount: BigInteger) {
        balanceMap = balanceMap.subtractAmountFromAddress(address, CryptoValue(cryptoCurrency, amount))
    }

    fun getAddressBalance(xpub: XPubs): CryptoValue =
        (getXpubBalance(xpub.forDerivation(XPub.Format.LEGACY)) +
            getXpubBalance(xpub.forDerivation(XPub.Format.SEGWIT))) as CryptoValue

    private fun getXpubBalance(xpub: XPub?): CryptoValue =
        xpub?.address?.let {
            return balanceMap[it]
        } ?: CryptoValue.zero(cryptoCurrency)

    fun updateAllBalances(
        xpubs: List<XPubs>,
        importedAddresses: List<String>
    ) {
        balanceMap = calculateCryptoBalanceMap(
            cryptoCurrency = cryptoCurrency,
            balanceQuery = balanceQuery,
            xpubs = xpubs,
            imported = importedAddresses
        )
    }

    @Deprecated("Use getBalanceQuery")
    abstract fun getBalanceOfAddresses(xpubs: List<XPubs>): Call<BalanceResponseDto>
}
